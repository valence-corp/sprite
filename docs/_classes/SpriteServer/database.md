---

layout: default

title: database

permalink: /classes/SpriteServer/database.html

---

### _SpriteServer_.database

#### Interface

(**databaseName: *string***)

Returns an SpriteDatabase client for the supplied `databaseName`,
using the authorization details of the `SpriteServer` client.

#### Example

```ts
const server = new SpriteServer({
  username: 'aUser',
  password: 'aPassword',
  address: 'http://localhost:2480',
});

const database = await server.database('aDatabase');
console.log(database.name);
// 'aDatabase';
```

