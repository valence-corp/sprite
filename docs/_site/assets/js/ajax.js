async function getDocument(url) {
  try {
    const response = await fetch(url);
    const html = await response.text();
    const parser = new DOMParser();
    const htmlDoc = parser.parseFromString(html, "text/html");
    const contentDiv = htmlDoc.getElementById("content");
    const contentHtml = contentDiv.outerHTML;
    document.getElementById("content").innerHTML = contentDiv.innerHTML;
  } catch (error) {
    throw new Error("Could not fetch page.", { cause: error });
  }
}

function createLinkClickEvents(root) {
  const elements = root.querySelectorAll("a");
  console.log(elements);
  if (elements.length) {
    elements.forEach((link) => {
      link.addEventListener("click", (e) => {
        e.preventDefault();
        getDocument(e.currentTarget.href);
      });
    });
  }
}

function ajax() {
  createLinkClickEvents(document);
}

document.addEventListener("DOMContentLoaded", ajax);
