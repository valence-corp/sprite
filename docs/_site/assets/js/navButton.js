document.addEventListener("DOMContentLoaded", function () {
  const navButton = document.getElementById("nav_button");
  const navIndicator = navButton.querySelector("svg");
  const mainNav = document.getElementById("mainNavigation");

  const closeIconMarkup = `
    <line x1="3" y1="4" x2="21" y2="21"></line>
    <line x1="21" y1="4" x2="3" y2="21"></line>
  `;

  const openIconMarkup = `
    <line x1="3" y1="4" x2="21" y2="4"></line>
    <line x1="3" y1="12" x2="21" y2="12"></line>
    <line x1="3" y1="20" x2="21" y2="20"></line>
  `;

  let isOpen = false;
  let previousTouchPosition = null;
  let currentTranslation = 0;

  function handleTouchEnd() {
    previousTouchPosition = null;

    if (currentTranslation > -80) {
      currentTranslation = 0;
      mainNav.style.transform = `translate3d(${currentTranslation}px, 0, 0)`;
    } else {
      mainNav.style.transform = "";
      currentTranslation = 0;
      closeMenu();
    }
  }

  function handleTouchStart(event) {
    previousTouchPosition = event.touches[0].clientX;
  }

  function handleTouchMove(event) {
    const currentTouchPosition = event.touches[0].clientX;
    const delta = currentTouchPosition - previousTouchPosition;
    previousTouchPosition = currentTouchPosition;

    currentTranslation += delta;

    if (currentTranslation > 0) {
      currentTranslation = 0;
    }

    mainNav.style.transform = `translate3d(${currentTranslation}px, 0, 0)`;
  }

  function enableTouchEvents() {
    mainNav.addEventListener("touchstart", handleTouchStart);
    mainNav.addEventListener("touchmove", handleTouchMove);
    mainNav.addEventListener("touchend", handleTouchEnd);
  }

  function disableTouchEvents() {
    mainNav.removeEventListener("touchstart", handleTouchStart);
    mainNav.removeEventListener("touchmove", handleTouchMove);
    mainNav.removeEventListener("touchend", handleTouchEnd);
  }

  function closeMenu() {
    toggleNavState();
    disableTouchEvents();
    navIndicator.innerHTML = openIconMarkup;
  }

  function openMenu() {
    toggleNavState();
    enableTouchEvents();
    navIndicator.innerHTML = closeIconMarkup;
  }

  function toggleNavState() {
    mainNav.classList.toggle("open");
    isOpen = !isOpen;
  }

  function handleNavButtonClick() {
    if (isOpen) {
      closeMenu();
    } else {
      openMenu();
    }
  }

  navButton.addEventListener("click", handleNavButtonClick);
});
