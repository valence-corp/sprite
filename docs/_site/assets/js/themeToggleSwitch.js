// function setTheme(theme) {
//   localStorage.setItem("theme", theme);
//   document.body.className = theme;
// }

// function setSwitchPosition(theme) {
//   switch (theme) {
//     case "dark":
//       themeToggleSwitch.checked = true;
//       break;
//     case "light":
//       themeToggleSwitch.checked = false;
//       break;
//     default:
//       throw new Error("unknow theme");
//   }
// }

// function setThemeAndSwitch(theme) {
//   setTheme(theme);
//   setSwitchPosition(theme);
// }

// function prefersTheme(theme) {
//   return window.matchMedia(`(prefers-color-scheme: ${theme})`).matches();
// }

// function initializeThemeToggle() {
//   const theme = localStorage.getItem("theme");
//   const toggleSwitch = document.getElementById("themeToggle");

//   toggleSwitch.addEventListener("click", () => {
//     if (themeToggleSwitch.checked) {
//       setTheme("dark");
//     } else {
//       setTheme("light");
//     }
//   });

//   if (theme) {
//     setThemeAndSwitch(theme);
//   } else {
//     if (prefersTheme("dark")) {
//       setThemeAndSwitch("dark");
//       return;
//     } else {
//       setThemeAndSwitch("light");
//     }
//   }
// }

// document.addEventListener("DOMContentLoaded", initializeThemeToggle);

class ThemeController {
  constructor(target) {
    this.element = target;
    this.theme = localStorage.getItem("theme");
    this.initialize();
  }
  initialize = () => {
    this.element.addEventListener("click", () => {
      if (this.element.checked) {
        this.changeTheme("dark");
      } else {
        this.changeTheme("light");
      }
    });
    this.determineInitialPosition();
  };
  prefersTheme = (theme) => {
    return window.matchMedia(`(prefers-color-scheme: ${theme})`).matches();
  };
  determineInitialPosition = () => {
    if (this.theme) {
      this.setTheme(this.theme);
    } else {
      if (prefersTheme("dark")) {
        this.setTheme("dark");
        return;
      } else {
        this.setTheme("light");
      }
    }
  };
  setPosition = (theme) => {
    switch (theme) {
      case "dark":
        this.element.checked = true;
        break;
      case "light":
        this.element.checked = false;
        break;
      default:
        throw new Error("unknow theme");
    }
  };
  changeTheme = (theme) => {
    localStorage.setItem("theme", theme);
    document.body.className = theme;
  };
  setTheme = (theme) => {
    this.changeTheme(theme);
    this.setPosition(theme);
  };
}

function initializeThemeToggle() {
  const element = document.getElementById("themeToggle");
  return new ThemeController(element);
}

document.addEventListener("DOMContentLoaded", initializeThemeToggle);
