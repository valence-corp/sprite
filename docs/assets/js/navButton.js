document.addEventListener("DOMContentLoaded", function () {
  const button = document.getElementById("navButton");
  const desktopButton = document.getElementById("navToggleDesktop");
  const indicator = button.querySelector("svg");
  const nav = document.getElementById("mainNavigation");
  let open = false;

  desktopButton.addEventListener("click", function () {
    nav.classList.toggle("open");
  });

  button.addEventListener("click", function () {
    nav.classList.toggle("open");
    open = !open;
    if (open) {
      indicator.innerHTML = `
        <line x1="34" y1="18" x2="18" y2="34"></line>
        <line x1="18" y1="18" x2="34" y2="34"></line>
      `;
    } else {
      indicator.innerHTML = `
        <line x1="34" y1="30" x2="18" y2="30"></line>
        <line x1="18" y1="22" x2="34" y2="22"></line>
      `;
    }
  });
});
