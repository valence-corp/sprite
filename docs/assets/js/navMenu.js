document.addEventListener("DOMContentLoaded", function () {
  const mainNav = document.getElementById("main_nav");
  const navButton = document.getElementById("nav_button");
  const navIndicator = navButton.querySelector("svg");

  const closeIconMarkup = `
    <line x1="3" y1="4" x2="21" y2="21"></line>
    <line x1="21" y1="4" x2="3" y2="21"></line>
  `;

  const openIconMarkup = `
    <line x1="3" y1="4" x2="21" y2="4"></line>
    <line x1="3" y1="12" x2="21" y2="12"></line>
    <line x1="3" y1="20" x2="21" y2="20"></line>
  `;

  let isOpen = false;
  let currentTranslation = 0;
  let previousTouchPosition = null;

  function handleTouchEnd() {
    previousTouchPosition = null;

    if (currentTranslation < -80) {
      closeMenu();
    }
    currentTranslation = 0;
    mainNav.style.transform = "";
  }

  function handleTouchStart(event) {
    previousTouchPosition = event.touches[0].clientX;
  }

  function handleTouchMove(event) {
    const currentTouchPosition = event.touches[0].clientX;
    const delta = currentTouchPosition - previousTouchPosition;
    previousTouchPosition = currentTouchPosition;

    currentTranslation += delta;

    if (currentTranslation > 0) {
      currentTranslation = 0;
    }

    mainNav.style.transform = `translate3d(${currentTranslation}px, 0, 0)`;
  }

  function enableTouchEvents() {
    mainNav.addEventListener("touchstart", handleTouchStart);
    mainNav.addEventListener("touchmove", handleTouchMove);
    mainNav.addEventListener("touchend", handleTouchEnd);
  }

  function disableTouchEvents() {
    mainNav.removeEventListener("touchstart", handleTouchStart);
    mainNav.removeEventListener("touchmove", handleTouchMove);
    mainNav.removeEventListener("touchend", handleTouchEnd);
  }

  function closeMenu() {
    toggleNavState();
    disableTouchEvents();
    navIndicator.innerHTML = openIconMarkup;
  }

  function openMenu() {
    toggleNavState();
    enableTouchEvents();
    navIndicator.innerHTML = closeIconMarkup;
  }

  function toggleNavState() {
    mainNav.classList.toggle("open");
    isOpen = !isOpen;
  }

  function handleNavButtonClick() {
    if (isOpen) {
      closeMenu();
    } else {
      openMenu();
    }
  }

  navButton.addEventListener("click", handleNavButtonClick);
});
