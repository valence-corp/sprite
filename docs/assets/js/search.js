"use strict";

import data from "/sprite/assets/js/searchData.js";
import serializedIndex from "./serialized.js";

async function initializeSearch() {
  const config = {
    minSearchLength: 3,
    indexedFields: ["name", "desc", "type", "parent"],
  };

  // Build the index using Lunr.js
  const index = lunr.Index.load(serializedIndex);

  // Define the root element
  const root = document.getElementById("searchBox");

  // State definition
  const state = {
    filters: {
      class: true,
      method: true,
      type: true,
    },
    results: {
      filtered: [],
      full: [],
    },
    visible: false,
  };

  // Elements for reference
  const el = {
    input: root.querySelector("#searchInput"),
    results: root.querySelector("#searchResultsBox"),
    list: root.querySelector("#resultList"),
    numberOfResults: root.querySelector("#numberOfResults"),
    numberResultsShowing: root.querySelector("#numberResultsShowing"),
    backdrop: undefined,
    filters: root.querySelector("#searchFilters").getElementsByTagName("input"),
    searchLabel: root.querySelector("#searchLabel"),
    close: root.querySelector("#searchClose"),
  };

  // Initialize search filter toggles
  for (let i = 0; i < el.filters.length; ++i) {
    el.filters[i].addEventListener("change", () => {
      state.filters[el.filters[i].value] = el.filters[i].checked;
      updateResultList();
    });
  }

  function appendBackdrop() {
    el.backdrop = document.createElement("div");
    el.backdrop.id = "backdrop";
    el.backdrop.addEventListener("click", hideSearchBox);
    root.appendChild(el.backdrop);
  }

  function removeBackdrop() {
    el.backdrop.removeEventListener("click", appendBackdrop);
    el.backdrop.remove();
    el.backdrop = undefined;
  }

  function handleKeyboardSelection() {
    const items = el.list.querySelectorAll("a");
    if (items.length > 0) {
      if (document.activeElement === el.input) {
        e.preventDefault();
        items[0].focus();
      } else {
        for (let i = 0; i < results.length; i++) {
          if (items[i] === document.activeElement) {
            e.preventDefault();
            if (i + 1 < items.length) {
              items[i + 1].focus();
            } else {
              el.input.focus();
            }
            break;
          }
        }
      }
    }
  }

  function checkKeyPress(e) {
    if (e.key === "Tab") {
      handleKeyboardSelection();
    }
    if (e.key === "Escape") {
      hideSearchBox();
    }
  }

  function showResultsBox() {
    appendBackdrop();
    state.visible = true;
    el.results.classList.remove("hidden");
    document.addEventListener("keydown", checkKeyPress);
  }

  function hideSearchBox() {
    state.visible = false;
    el.results.classList.add("hidden");
    document.removeEventListener("keydown", showResultsBox);
    removeBackdrop();
  }

  function updateResultList() {
    state.results.filtered = [];
    el.list.innerHTML = "";
    el.numberOfResults.innerHTML = state.results.full.length;
    el.numberResultsShowing.innerHTML = state.results.filtered.length;

    if (state.results.full.length > 0) {
      state.results.full.forEach((result) => {
        // TODO: should look into this, is it necessary to find
        // the original document?
        const doc = data.find((doc) => doc.name === result.ref);
        const type = doc.type === "interface" ? "type" : doc.type;
        if (state.filters[type]) {
          state.results.filtered.push(result.ref);
          const section = document.createElement("section");
          const desc = document.createElement("p");
          const link = document.createElement("a");
          const li = document.createElement("li");

          li.className = doc.type;

          link.href = `/sprite/doc/${doc.name}`;
          link.className = "searchResultLink";
          link.addEventListener("click", () => {});
          link.textContent = doc.parent
            ? `${doc.parent}.${doc.name}`
            : doc.name;

          desc.textContent = doc.desc;

          section.appendChild(link);
          section.appendChild(desc);

          li.appendChild(section);

          el.list.appendChild(li);
        }
      });

      el.numberResultsShowing.innerHTML = state.results.filtered.length;
    }
  }

  function displayResults() {
    if (!state.visible) {
      showResultsBox();
    }
    updateResultList();
  }

  // search the index
  function search() {
    // Hide the "Search" Label
    // (which functions as a placeholder)
    if (el.input.value.length > 0) {
      el.searchLabel.classList.add("hidden");
    } else {
      el.searchLabel.classList.remove("hidden");
    }

    // run a search if the search term is longer than
    // the minimum set in the config
    if (el.input.value.length >= config.minSearchLength) {
      const initial = index.search(searchInput.value);

      // if there are no results, try again with a wildcard,
      // and then yet again with two wildcards
      if (initial.length === 0) {
        const wildcardResults = index.search(`${searchInput.value}*`);
        state.results.full =
          wildcardResults.length > 0
            ? wildcardResults
            : index.search(`*${searchInput.value}*`);
      } else {
        state.results.full = initial;
      }

      displayResults();
    }
  }

  el.input.addEventListener("input", search);
  el.close.addEventListener("click", hideSearchBox);
}

document.addEventListener("DOMContentLoaded", initializeSearch);
