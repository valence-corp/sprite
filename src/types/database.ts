import { ArcadeSelectTimeoutStrategy } from '../nodes/types.js';
import { ISpriteRestClientConnectionParameters } from './client.js';
import { SpriteRestClient } from '../SpriteRestClient.js';
import { SpriteTransaction } from '../SpriteTransaction.js';
import { ValidSuperTypeKey } from '../SpriteType.js';

export type ArcadeRecordCategory = 'd' | 'e' | 'v';
export type ArcadeRecordType = 'document' | 'edge' | 'vertex';

export type RecordBase = {
  '@rid': string;
  '@cat': ArcadeRecordCategory;
  '@type': string;
};

export type EdgeBase = RecordBase & {
  '@in': string;
  '@out': string;
};

export type WithArcadeEdgeRecordMeta<S> = {
  [K in keyof S]: S[K] & EdgeBase;
};

export type WithArcadeRecordMeta<S> = {
  [K in keyof S]: S[K] & RecordBase;
};

export type TypeNames<Schema> = keyof Schema;

/**
 * The parameters necessary to perform opertations
 * on a target ArcadeDB database.
 */
export interface ISpriteDatabaseConnectionParameters
  extends ISpriteRestClientConnectionParameters {
  /**
   * The name of the database to connect to.
   */
  databaseName: string;
}

export interface ISpriteDatabaseClientParameters {
  client: SpriteRestClient;
  databaseName: string;
}

/**
 * The Query languages supported by ArcadeDB, to be declared when
 * using SpriteDatabase.query() and SpriteDatabase.command() methods
 */
export type ArcadeSupportedQueryLanguages =
  | 'sql'
  | 'sqlscript'
  | 'graphql'
  | 'cypher'
  | 'gremlin'
  | 'mongo';

/**
 * Options to create a new type with.
 */
export interface ISpriteCreateTypeOptions<S, N extends TypeNames<S>> {
  /**
   * When set to true, the type creation will be ignored if the
   * type already exists (instead of failing with an error).
   * @default false
   */
  ifNotExists?: boolean;
  /**
   * Defines a super-type you want to extend with this type.
   */
  extends?: ValidSuperTypeKey<S, N>;
  /**
   * A bucket-name, or an array of bucket-names you want this type to use.
   */
  buckets?: string | Array<string>;
  /**
   * Defines the total number of buckets you want to create for this type. The
   * @default 1
   */
  totalBuckets?: number;
}

export type OmitMeta<T> = Omit<T, keyof RecordBase | keyof EdgeBase>;

/** Options to insert a new record with */
export interface ISpriteInsertRecordOptions<T> {
  /** The data to populate the newly created reacord with */
  data?: OmitMeta<T>;
  /** The bucket to store the record in */
  bucket?: string;
}

/**
 * Options for a database.dropType() command
 */
export interface ISpriteDropTypeOptions {
  /**
   * Prevent errors if the type does not exits when attempting to drop it.
   * @default false;
   */
  ifExists?: boolean;
  /**
   * Defines whether the command drops non-empty edge and vertex types. Note, this can
   * disrupt data consistency. Be sure to create a backup before running it.
   * @default: false
   */
  unsafe?: boolean;
}

/** A function which contains the operations to be conducted within a transaction */
export type SpriteTransactionCallback =
  | ((trx: SpriteTransaction) => Promise<void>)
  | ((trx: SpriteTransaction) => void);

/**
 * How the bucket will be selected for a newly created record of this type
 * @default 'round-robin'
 */
export type ArcadeBucketSelectionStrategies =
  | 'round-robin'
  | 'thread'
  | 'partitioned<primary-key>';

/**
 * A type definition returned by ArcadeDB when a getSchema command
 * is compeleted.
 */
export type ArcadeTypeDefinition = {
  /** The name of the type */
  name: string;
  /** The category of the type (document, vertex, or edge) */
  type: ArcadeRecordType;
  /**
   * The number of records with this type name
   */
  records: number;
  /**
   * The buckets associated with this type.
   */
  buckets: Array<string>;
  /** How the bucket will be selected for a newly created record of this type */
  bucketSelectionStrategy: ArcadeBucketSelectionStrategies;
  /**
   * Super types associated with this type.
   */
  parentTypes: Array<string>;
  /**
   * Properties defined for this type.
   */
  properties: Array<string>;
  /** Indexes defined for records of this type */
  indexes: Array<string>;
  /** An object for custom user-defined properties for this type */
  custom: Record<string, unknown>;
};

export type ArcadeGetSchemaResponse = Array<ArcadeTypeDefinition>;

export type ArcadeSqlExplanationExecutionPlan = {
  type: 'QueryExecutionPlan' | 'DDLExecutionPlan';
  javaType: string;
  cost: number;
  prettyPrint: string;
  steps: object[];
};

/**
 * An object representing the explanation of an
 * SQL command.
 */
export type ArcadeSqlExplanation = {
  executionPlan: ArcadeSqlExplanationExecutionPlan;
  executionPlanAsString: string;
};

/**
 * An object returned from the server following a successfully command / query
 *
 */
export interface ArcadeCommandResponse<T = unknown> {
  /**
   * The user that issued the command / query.
   */
  user: string;
  /** The version of ArcadeDB returning the result. */
  version: string;
  /**
   * The name of the server returning the result.
   */
  serverName: string;
  /**
   * The result of the command / query.
   */
  result: T;
}

export type ArcadeQueryResponse<T = unknown> = ArcadeCommandResponse<T>;

// TODO: I have not checked these for compatibility with ArcadeDB
export declare const COMPARISON_OPERATORS: readonly [
  '=',
  '==',
  '!=',
  '<>',
  '>',
  '>=',
  '<',
  '<=',
  'in',
  'not in',
  'is',
  'is not',
  'like',
  'not like',
  'match',
  'ilike',
  'not ilike',
  '@>',
  '<@',
  '&&',
  '?',
  '?&',
  '!<',
  '!>',
  '<=>',
  '!~',
  '~',
  '~*',
  '!~*',
  '@@',
  '@@@',
  '!!',
  '<->',
  'regexp',
  'is distinct from',
  'is not distinct from',
];

export type SpriteOperators = (typeof COMPARISON_OPERATORS)[number];

export interface ISpriteSelectFromOptions<
  S,
  N extends keyof S,
  P extends keyof WithRid<S, N>,
> {
  /**
   * Designates conditions to filter the result-set.
   */
  where?: SpriteWhereClause<S, N, P>;
  /**
   * Designates the field with which to order the result-set.
   * Use the optional 'ASC' and 'DESC' operators to define the direction of the order.
   */
  orderBy?: {
    field: keyof S[N];
    /**
     * Defines the direction to sort the result (ASCending or DESCending).
     * @default 'ASC'.
     */
    direction?: 'ASC' | 'DESC';
  };
  /**
   * Defines the number of records you want to skip from the start of the result-set.
   * You mayfind this useful in Pagination, when using it in conjunction with the
   * limit `option`.
   */
  skip?: number;
  /**
   * Defines the maximum number of records in the result-set. You may find this useful in
   * Pagination, when using it in conjunction with the `skip` option.
   */
  limit?: number;
  /**
   * Defines the maximum time in milliseconds for the query, and optionally the
   * exception strategy to use.
   */
  timeout?: {
    /**
     *  The duration of the timeout in milliseconds.
     */
    duration: number;
    /**
     * The timeout strategy to use.\
     * `RETURN` Truncates the result-set, returning the data collected up to the timeout.\
     * `EXCEPTION` Raises an exception.
     */
    strategy?: ArcadeSelectTimeoutStrategy;
  };
}

export type WithRid<S, N extends TypeNames<S>> = OmitMeta<S[N]> & {
  '@rid': string;
};

export type SpriteWhereClause<
  S,
  N extends TypeNames<S>,
  P extends keyof WithRid<S, N>,
> = [P, SpriteOperators, WithRid<S, N>[P]];

export interface ISpriteDeleteFromOptions<
  S,
  N extends TypeNames<S>,
  P extends keyof WithRid<S, N>,
> {
  /**
   * Designates conditions to filter the result-set.
   */
  where: SpriteWhereClause<S, N, P>;
  /**
   * The duration of the timeout in milliseconds.
   */
  timeout?: number;
  /**
   * Defines the maximum number of records in the result-set.
   * @default undefined
   */
  limit?: number;
  /**
   * Defines what is returned following the command: the count of the records before (`BEFORE`) or following deletion (`COUNT`).
   * @default 'COUNT'
   */
  return?: 'COUNT' | 'BEFORE';
}
