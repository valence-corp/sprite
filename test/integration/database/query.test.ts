import { ArcadeSupportedQueryLanguages } from '../../../src/types/database.js';
import { testClient as client } from './testClient.js';

describe('SpriteDatabase.query', () => {
  it('executes a query successfully', async () => {
    const command = 'SELECT FROM schema:types';
    const language: ArcadeSupportedQueryLanguages = 'sql';
    const result = await client.query(language, command);

    expect(result).toMatchSnapshot();
  });

  it('propagates errors from ArcadeDB', async () => {
    const command = 'INVALID COMMAND';
    const language = 'INVALID_LANGUAGE';
    await expect(
      // @ts-expect-error - Testing invalid input
      client.query(language, command)
    ).rejects.toMatchSnapshot();
  });
});
