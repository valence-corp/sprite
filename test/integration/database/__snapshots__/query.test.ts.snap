// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpriteDatabase.query executes a query successfully 1`] = `
[
  {
    "bucketSelectionStrategy": "round-robin",
    "buckets": [
      "Flavour_0",
      "Flavour_1",
      "Flavour_2",
      "Flavour_3",
      "Flavour_4",
      "Flavour_5",
      "Flavour_6",
      "Flavour_7",
    ],
    "custom": {},
    "indexes": [],
    "name": "Flavour",
    "parentTypes": [],
    "properties": [],
    "records": 24,
    "type": "vertex",
  },
  {
    "bucketSelectionStrategy": "round-robin",
    "buckets": [
      "FlavourGroup_0",
      "FlavourGroup_1",
      "FlavourGroup_2",
      "FlavourGroup_3",
      "FlavourGroup_4",
      "FlavourGroup_5",
      "FlavourGroup_6",
      "FlavourGroup_7",
    ],
    "custom": {},
    "indexes": [],
    "name": "FlavourGroup",
    "parentTypes": [],
    "properties": [],
    "records": 20,
    "type": "vertex",
  },
  {
    "bucketSelectionStrategy": "round-robin",
    "buckets": [
      "FlavourInGroup_0",
      "FlavourInGroup_1",
      "FlavourInGroup_2",
      "FlavourInGroup_3",
      "FlavourInGroup_4",
      "FlavourInGroup_5",
      "FlavourInGroup_6",
      "FlavourInGroup_7",
    ],
    "custom": {},
    "indexes": [],
    "name": "FlavourInGroup",
    "parentTypes": [],
    "properties": [],
    "records": 48,
    "type": "edge",
  },
  {
    "bucketSelectionStrategy": "round-robin",
    "buckets": [
      "HasFlavours_0",
      "HasFlavours_1",
      "HasFlavours_2",
      "HasFlavours_3",
      "HasFlavours_4",
      "HasFlavours_5",
      "HasFlavours_6",
      "HasFlavours_7",
    ],
    "custom": {},
    "indexes": [],
    "name": "HasFlavours",
    "parentTypes": [],
    "properties": [],
    "records": 72,
    "type": "edge",
  },
  {
    "bucketSelectionStrategy": "round-robin",
    "buckets": [
      "InRegion_0",
      "InRegion_1",
      "InRegion_2",
      "InRegion_3",
      "InRegion_4",
      "InRegion_5",
      "InRegion_6",
      "InRegion_7",
    ],
    "custom": {},
    "indexes": [],
    "name": "InRegion",
    "parentTypes": [],
    "properties": [],
    "records": 45,
    "type": "edge",
  },
  {
    "bucketSelectionStrategy": "round-robin",
    "buckets": [
      "Location_0",
      "Location_1",
      "Location_2",
      "Location_3",
      "Location_4",
      "Location_5",
      "Location_6",
      "Location_7",
    ],
    "custom": {},
    "indexes": [],
    "name": "Location",
    "parentTypes": [],
    "properties": [],
    "records": 90,
    "type": "vertex",
  },
  {
    "bucketSelectionStrategy": "round-robin",
    "buckets": [
      "ORIDs_0",
      "ORIDs_1",
      "ORIDs_2",
      "ORIDs_3",
      "ORIDs_4",
      "ORIDs_5",
      "ORIDs_6",
      "ORIDs_7",
    ],
    "custom": {},
    "indexes": [],
    "name": "ORIDs",
    "parentTypes": [],
    "properties": [],
    "records": 1,
    "type": "document",
  },
  {
    "bucketSelectionStrategy": "round-robin",
    "buckets": [
      "ProducedIn_0",
      "ProducedIn_1",
      "ProducedIn_2",
      "ProducedIn_3",
      "ProducedIn_4",
      "ProducedIn_5",
      "ProducedIn_6",
      "ProducedIn_7",
    ],
    "custom": {},
    "indexes": [],
    "name": "ProducedIn",
    "parentTypes": [],
    "properties": [],
    "records": 72,
    "type": "edge",
  },
  {
    "bucketSelectionStrategy": "round-robin",
    "buckets": [
      "Region_0",
      "Region_1",
      "Region_2",
      "Region_3",
      "Region_4",
      "Region_5",
      "Region_6",
      "Region_7",
    ],
    "custom": {},
    "indexes": [],
    "name": "Region",
    "parentTypes": [],
    "properties": [],
    "records": 12,
    "type": "vertex",
  },
  {
    "bucketSelectionStrategy": "round-robin",
    "buckets": [
      "Whisky_0",
      "Whisky_1",
      "Whisky_2",
      "Whisky_3",
      "Whisky_4",
      "Whisky_5",
      "Whisky_6",
      "Whisky_7",
    ],
    "custom": {},
    "indexes": [],
    "name": "Whisky",
    "parentTypes": [],
    "properties": [],
    "records": 144,
    "type": "vertex",
  },
]
`;

exports[`SpriteDatabase.query propagates errors from ArcadeDB 1`] = `[Error: Invalid language or query. Status: 400]`;
